### ISSUE TITLE
Run Travis-CI in OSX in addition to Ubuntu

### ISSUE THREAD

**Issue by gerudo7** [villano-lab/nrCascadeSim#29]
It would be nice to have some sort of automatic testing for OSX to help with troubleshooting for Macs. Here is the [documentation](https://docs.travis-ci.com/user/reference/osx) for OSX testing through Travis-CI, and here is the [documentation](https://docs.travis-ci.com/user/multi-os/) for testing on multiple operating systems. It looks like including testing for both should be pretty straightforward.

**Comment by villaa** [comment]
check the OSX distributions on travis [here](https://docs.travis-ci.com/user/reference/osx/).

**Comment by villaa** [comment]
Check out the techniques used [here](https://docs.travis-ci.com/user/multi-os/) for mac OSX builds. And also perhaps [here](https://stackoverflow.com/questions/45257534/how-can-i-build-a-python-project-with-osx-environment-on-travis).

**Comment by villaa** [comment]
Installing [miniconda from the command line](https://docs.conda.io/projects/conda/en/latest/user-guide/install/macos.html) might be a solution to getting a reasonable python environment going on OSX.

**Comment by villaa** [comment]
WARNING:[ job 621 passed ](https://app.travis-ci.com/github/villano-lab/nrCascadeSim/builds/243908744)the CI but it looks like `realizeCascades` *did not* run properly.  Why??

**Comment by villaa** [comment]
> WARNING:[ job 621 passed ](https://app.travis-ci.com/github/villano-lab/nrCascadeSim/builds/243908744)the CI but it looks like `realizeCascades` _did not_ run properly. Why??

When ROOT output file isn't created at all it seems to pass!!?, that's a bug.

**Comment by villaa** [comment]
[This](https://root-forum.cern.ch/t/rpath-madness-on-os-x/23654/2) error comes up on OSX builds. Tried to use the work-around listed in the post.

**Comment by gerudo7** [comment]
> When ROOT output file isn't created at all it seems to pass!!?, that's a bug.

Shoot. I think this is an oversight on my part. If I run the python script directly I definitely get an error code if the python script fails. Working on a fix now.

**Comment by gerudo7** [comment]
@villaa Okay, just fixed this on develop.

> When ROOT output file isn't created at all it seems to pass!!?, that's a bug.

**Comment by villaa** [comment]
I think I now have the OSX compile "working" but still getting a segfault error (only on OSX): ` [/usr/lib/system/libsystem_platform.dylib] _sigtramp (no debug info)` 

Trying to debug this but [this](https://github.com/micahfolsom/qtcanvas/issues/1) thread suggests an X11 issue and the [linked issue](https://sft.its.cern.ch/jira/browse/ROOT-8886) suggests it could be because ROOT itself or Qt were not compiled with X11 (XQuartz). 

I don't know what's with this problem because I'm not even using graphics at all, so why should all this stuff matter?

**Comment by villaa** [comment]
Still having  some odd segfaults for OSX. Time to bring out the big guns: `gdb`. I installed with `brew` like `brew install gdb`

What ensued was basically a 4 hour code-signing debacle because Mac is annoying. In any case [these instructions](https://superuser.com/questions/1436370/how-to-codesign-gdb-on-os-x-mojave)  eventually worked. Yes, TWO reboots were required.

**Comment by villaa** [comment]
Update: even after installing and signing `gdb` am having a lot of trouble correctly running it.  When I finally  get the executable correctly instrumented (I used `-g -O0` flag and ran `gdb` on the original executable NOT the one made with a `sudo make` command) the `gdb` program hangs with something like:

```
Note: this version of macOS has System Integrity Protection.
Because `startup-with-shell' is enabled, gdb has worked around this by
caching a copy of your shell.  The shell used by "run" is now:
    /Users/villaa/Library/Caches/gdb/bin/sh
[New Thread 0x1703 of process 3447]
[New Thread 0x2803 of process 3447]
```

Instead I'm using `lldb` because of [this snarky post](https://discussions.apple.com/thread/252960654).

**Comment by villaa** [comment]
while debugging found that the regex library I use to read the input files seems to be failing on OSX, and the behavior is different on Linux. 

the library include is: `#include<regex.h>` and the object I'm using is in `src/cascadeProd.c` and is of type `regex_t` and the function `regcomp` used for the matching is not producing any  matches on OSX (although it is on Linux).

**Comment by villaa** [comment]
[this](https://regex101.com/r/5tJpvb/1)  is the regular expression i'm playing with to see why it evaluates differently on OSX (Mojave).

**Comment by villaa** [comment]
On OSX (Mojave) the regex function `regcomp()` fails with this code: `regexec() failed with 'repetition-operator operand invalid'`

**Comment by villaa** [comment]
Ugh, might I have to [use boost](https://www.boost.org/doc/libs/1_64_0/libs/regex/doc/html/boost_regex/syntax/basic_extended.html)?

**Comment by villaa** [comment]
Ok, so I think I just have to avoid using `+?` and `*?` as regex operators. I think I was using them incorrectly in the first regex quoted above anyway. They are meant to be taken together to give "lazy" action and I was using the `?` to mean zero or one times for an expression like `[0-9]+?` by which I meant either no digits or any number, but really it was matching any number of digits "lazily." 

here is an update of my regex: [v2](https://regex101.com/r/NDmdY0/1)

**Comment by villaa** [comment]
here is an update of my regex: [v3](https://regex101.com/r/8g6U0G/1)

**Comment by villaa** [comment]
here is another update: [v4](https://regex101.com/r/RVj3SD/1)

**Comment by villaa** [comment]
final one that I'll try: [v5](https://regex101.com/r/NnLbzQ/1)

This one doesn't have any `*?` or `+?`

**Comment by villaa** [comment]
Making  some progress based on my [SO post](https://stackoverflow.com/questions/70488948/regex-h-matching-differences-between-osx-and-linux?noredirect=1#comment124605174_70488948), can probably start a pull  request. Regex is weird, at least as implemented in `regex.h`; I was having a lot of trouble with the `\s+` so one of the suggestions in my post showed me how to do  that a different way: `[[:space:]]`  supposedly this is more POSIX.

**Comment by villaa** [comment]
REALIZATION: even on Linux, the interpretation of scientific notation as double has always failed b/c the function `interpretDbl` in `cascadeProd.c` is not good at selecting scientific notation OR other double types.

**Comment by villaa** [comment]
Addressed with  PR #61
